### ISSUE TITLE
How to use app.stub.ts

### ISSUE THREAD

**Issue by DotHide** [lathonez/clicker#34]
Hey lathonez, thx for this sample project. It's cute & really helpful. What I don't know is how to use `app.stub.ts` .

I know it is a stub for the `@App` decorator, but I don't know how to test decorators. Could u pls give me a little prompt? thx~
> The reason why I want to test decorators is I want to get a full coverage report 100% ：）

**Comment by lathonez** [comment]
Hey, great question.

TL;dr, `app.stub.ts` is used by our setup to prevent a bunch of errors during test, it has no other purpose.

The _only_ reason I have app.stub.ts is,

When you remove it:

```
x220:~/code/clicker$ git diff
diff --git a/test/karma.config.js b/test/karma.config.js
index 42eecc9..4a76de3 100644
--- a/test/karma.config.js
+++ b/test/karma.config.js
@@ -35,7 +35,7 @@ module.exports = function(config) {
     exclude: [
       'node_modules/angular2/**/*_spec.js',
       'node_modules/ionic-angular/**/*spec*',
-      'node_modules/ionic-angular/decorators/app.js'
+      // 'node_modules/ionic-angular/decorators/app.js'
     ],
 
     // preprocess matching files before serving them to the browser
@@ -67,7 +67,7 @@ module.exports = function(config) {
     //           Also any files you want to serve need to be in the files array above with serverd: true
     proxies: {
       // allows us to keep test code separate from app code and still have the references work
-      '/base/node_modules/ionic-angular/decorators/app.js': '/base/www/build/test/app.stub.js', // stub out Ionic's @App decorator
+      // '/base/node_modules/ionic-angular/decorators/app.js': '/base/www/build/test/app.stub.js', // stub out Ionic's @App decorator
       '/base/ionic-angular.js': '/base/test/ionic-angular.js'
     },
```

and test:

```
x220:~/code/clicker$ npm test

> clicker@ test /home/lathonez/code/clicker
> gulp --gulpfile test/gulpfile.ts --cwd ./ test

[17:24:01] Requiring external module ts-node/register
[17:24:10] Using gulpfile ~/code/clicker/test/gulpfile.ts
[17:24:10] Starting 'test'...
[17:24:10] Starting 'test.clean'...
[17:24:10] Starting 'test.lint'...
[17:24:12] Deleted /home/lathonez/code/clicker/www/build/app.html, /home/lathonez/code/clicker/www/build/components, /home/lathonez/code/clicker/www/build/css, /home/lathonez/code/clicker/www/build/fonts, /home/lathonez/code/clicker/www/build/pages
[17:24:12] Finished 'test.clean' after 1.25 s
[17:24:15] Finished 'test.lint' after 4.27 s
[17:24:15] Starting 'test.build.html'...

∆ Copying HTML
√ Matching patterns: app/**/*.html
[17:24:15] Finished 'test.build.html' after 37 ms
[17:24:15] Starting 'test.build.fonts'...

∆ Copying fonts
√ Matching patterns: node_modules/ionic-angular/fonts/**/*.+(ttf|woff|woff2)
[17:24:15] Finished 'test.build.fonts' after 8.72 ms
[17:24:15] Starting 'test.build.sass'...

∆ Compiling Sass to CSS
√ Matching patterns: app/theme/app.+(ios|md).scss
[17:24:15] Finished 'test.build.sass' after 18 ms
[17:24:15] Starting 'test.build.locker'...
[17:24:15] buildLocker: waiting for fonts build to complete
√ HTML copied to www/build
√ Fonts copied to www/build/fonts
[17:24:15] buildLocker: waiting for css build to complete
[17:24:15] buildLocker: waiting for css build to complete
[17:24:15] buildLocker: waiting for css build to complete
[17:24:15] buildLocker: waiting for css build to complete
[17:24:17] buildLocker: waiting for css build to complete
[17:24:17] buildLocker: waiting for css build to complete
[17:24:17] buildLocker: waiting for css build to complete
[17:24:17] buildLocker: waiting for css build to complete
[17:24:18] buildLocker: waiting for css build to complete
√ Sass compilation complete
[17:24:18] buildLocker: unlocked
[17:24:18] Finished 'test.build.locker' after 3.38 s
[17:24:18] Starting 'test.build.typescript'...
[17:24:29] Finished 'test.build.typescript' after 10 s
[17:24:29] Starting 'test.karma'...

START:
06 03 2016 17:24:33.945:INFO [karma]: Karma v0.13.21 server started at http://localhost:9876/
06 03 2016 17:24:33.966:INFO [launcher]: Starting browser PhantomJS
06 03 2016 17:24:34.619:INFO [PhantomJS 2.1.1 (Linux 0.0.0)]: Connected on socket /#ODerMqY-CsOKeaXIAAAA with id 13630062
PhantomJS 2.1.1 (Linux 0.0.0) WARN: 'DEPRECATION WARNING: 'enqueueTask' is no longer supported and will be removed in next major release. Use addTask/addRepeatingTask/addMicroTask'
ERROR: 'EXCEPTION: The selector "ion-app" did not match any elements'
ERROR: 'STACKTRACE:'
ERROR: '

-----async gap-----
loadAsRoot
useFactory
_instantiate
_instantiateProvider
_new
getObjByKeyId
_getByKeyDefault
_getByKey
get


run
bootstrap
bootstrap

DecorateConstructor@http://localhost:9876/base/node_modules/reflect-metadata/Reflect.js?dda56e9eda58525388f2f12f0e03222eea4db7ec:536:38
decorate@http://localhost:9876/base/node_modules/reflect-metadata/Reflect.js?dda56e9eda58525388f2f12f0e03222eea4db7ec:98:39



eval code
eval@[native code]
__exec@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:1400:16
execute@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:3778:22
linkDynamicModule@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:3132:36
getModule@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:3100:26
http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:3136:25
require@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:3748:34

eval code
eval@[native code]
__exec@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:1400:16
execute@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:3778:22
linkDynamicModule@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:3132:36
link@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:2975:28
execute@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:3312:17
doDynamicExecute@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:708:32
link@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:910:36
doLink@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:562:11
updateLinkSetOnLoad@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:610:24
http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:422:30

-----async gap-----
_createInnerZone
NgZone
createNgZone
application
bootstrap

DecorateConstructor@http://localhost:9876/base/node_modules/reflect-metadata/Reflect.js?dda56e9eda58525388f2f12f0e03222eea4db7ec:536:38
decorate@http://localhost:9876/base/node_modules/reflect-metadata/Reflect.js?dda56e9eda58525388f2f12f0e03222eea4db7ec:98:39



eval code
eval@[native code]
__exec@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:1400:16
execute@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:3778:22
linkDynamicModule@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:3132:36
getModule@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:3100:26
http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:3136:25
require@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:3748:34

eval code
eval@[native code]
__exec@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:1400:16
execute@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:3778:22
linkDynamicModule@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:3132:36
link@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:2975:28
execute@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:3312:17
doDynamicExecute@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:708:32
link@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:910:36
doLink@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:562:11
updateLinkSetOnLoad@http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:610:24
http://localhost:9876/base/node_modules/systemjs/dist/system.src.js?281591da6090abed69cd0b080aa69c8de089a41c:422:30'
  Click
    ✔ initialises with defaults
    ✔ initialises with overrides
  Clicker
    ✔ initialises with the correct name
  Utils
    ✔ resets a control
  ClickerApp
    ✔ initialises with two possible pages
    ✔ initialises with a root page
    ✔ initialises with an app
    ✔ opens a page
  ClickerButton
    ✔ initialises
    ✔ displays the clicker name and count
    ✔ does a click
  ClickerForm
    ✔ initialises
WARN: 'DEPRECATION WARNING: 'dequeueTask' is no longer supported and will be removed in next major release. Use removeTask/removeRepeatingTask/removeMicroTask'
    ✔ passes new clicker through to service
    ✔ doesn't try to add a clicker with no name
  Clickers
    ✔ initialises with empty clickers
    ✔ creates an instance of SqlStorage
    ✔ has empty ids with no storage
    ✔ has empty clickers with no storage
    ✔ can initialise a clicker from string
    ✔ returns undefined for a bad id
    ✔ adds a new clicker with the correct name
    ✔ removes a clicker by id
    ✔ does a click
    ✔ loads IDs from storage
    ✔ loads clickers from storage
  ClickerList
    ✔ initialises

Finished in 0.834 secs / 0.832 secs

SUMMARY:
✔ 26 tests completed
--------------------------------|----------|----------|----------|----------|----------------|
File                            |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
--------------------------------|----------|----------|----------|----------|----------------|
 test/                          |    89.74 |    48.39 |       80 |    94.29 |                |
  app.js                        |    89.66 |    48.28 |    85.71 |       96 |              5 |
  testUtils.js                  |       90 |       50 |    66.67 |       90 |              8 |
 test/components/clickerButton/ |       85 |    48.28 |      100 |    93.75 |                |
  clickerButton.js              |       85 |    48.28 |      100 |    93.75 |              5 |
 test/components/clickerForm/   |    90.32 |    51.61 |      100 |     96.3 |                |
  clickerForm.js                |    90.32 |    51.61 |      100 |     96.3 |              5 |
 test/models/                   |      100 |      100 |      100 |      100 |                |
  click.js                      |      100 |      100 |      100 |      100 |                |
  clicker.js                    |      100 |      100 |      100 |      100 |                |
 test/pages/clickerList/        |    86.96 |    48.28 |      100 |    94.74 |                |
  clickerList.js                |    86.96 |    48.28 |      100 |    94.74 |              5 |
 test/pages/page2/              |    82.35 |    48.28 |       75 |    92.31 |                |
  page2.js                      |    82.35 |    48.28 |       75 |    92.31 |              5 |
 test/services/                 |    92.55 |    48.39 |    86.67 |    95.35 |                |
  clickers.js                   |    94.94 |    48.39 |       96 |    97.22 |           5,67 |
  utils.js                      |       80 |      100 |       40 |    85.71 |            6,8 |
--------------------------------|----------|----------|----------|----------|----------------|
All files                       |    90.87 |       50 |    89.71 |    95.54 |                |
--------------------------------|----------|----------|----------|----------|----------------|
```

You get loads of errors (^^^). The only way I could remove the errors was to stub out the decorator.

Currently [clickerForm.spec.ts](https://github.com/lathonez/clicker/blob/master/app/components/clickerForm/clickerForm.spec.ts) is as far as I've gone in terms of deep angular testing. This tests a @component decorator by interacting with the DOM.

I've been meaning to try it with @page too, I think it'll be the same set up.

**Comment by lathonez** [comment]
Please re-open if necessary :)

**Comment by DotHide** [comment]
Thx. Now I know the purpose of `app.stub.ts` . You just use it instead of the original one, cause PhantomJS doesn't know the element `<ion-app></ion-app>`. Is it?

So, the further question is **Must I fake all of element prefixed with `ion-` _only_ by this way?**

1. exclude it
2. proxy it
3. to avoid `blow up` messages ：）

Another one is about the deep angular testing. I mentioned the `Branch` is always less than 50% when there're **Decorators** in a Class. If I delete them, the `Branch` will become 100%:

```ts
...
// @App({
//   templateUrl: 'build/app.html',
//   config: {}, // http://ionicframework.com/docs/v2/api/config/Config/
// })
...
```
Okey, I know it's not the correct way to solve problems. 
So, the question is why Branch point of `clickerForm.js` is still only 51.61% after the deep angular testing and how to improve it?

Thx.

**Comment by lathonez** [comment]
Another really good question.

> So, the further question is Must I fake all of element prefixed with ion- only by this way?

Absolutely not. The point of [this code](https://github.com/lathonez/clicker/blob/master/app/components/clickerForm/clickerForm.spec.ts#L43-L49) is to provide the Ionic elements to your tests; this an example of testing an [`<ion-input..`](https://github.com/lathonez/clicker/blob/master/app/components/clickerForm/clickerForm.html#L5)

I found [this guide](https://developers.livechatinc.com/blog/testing-angular-2-apps-dependency-injection-and-components/) really helpful when I first wrote this stuff.

Now you have said it, I think concept this should work for `<ion-app` and negate the purpose of `app.stub.ts`. I will have a play with it now and see if I can remove it.

**Comment by lathonez** [comment]
Unfortunately not, the @App decorator / `<ion-app` seems to be a special case. I am able to test it properly using the dependency injection method in clickerForm.spec.ts and that guide I referenced, however I still get the same errors in totally unrelated parts of the test cycle.

So - it is necessary to stub it out for now. If you manage to find out the underlying reason and remove the stub please let me know.

This only applies to `<ion-app`, not any other ionic decorators that I've come across. They can be fiddly to test, that guide and clickerForm.spec.ts are good reference points.

**Comment by lathonez** [comment]
Regarding the test coverage, the only uncovered line I have is:

```javascript
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
```

This is the same in all files. As it's transpiled code that I have no control over I have never bothered with it. I just make sure all my core logic is covered. This amounts to just having one uncovered line in all files (so if I see more than one, I know I need to sort out my coverage).

Again, if you find out how to reach this line in tests, I would be keen to know how.

**Comment by srishiva89** [comment]
When i run my unit test cases, I'm getting below error:-

PhantomJS 2.1.1 (Windows 8 0.0.0) ERROR: 'ORIGINAL EXCEPTION: The selector "ion-app" did not match any elements'
PhantomJS 2.1.1 (Windows 8 0.0.0) ERROR: 'ERROR CONTEXT:'
PhantomJS 2.1.1 (Windows 8 0.0.0) ERROR: DebugContext{_view: DebugAppView{clazz: function _View_DxApp_Host0(viewUtils, parentInjector, declarationEl) { ... }, componentType: RenderComponentType{id: ..., templateUrl: ..., slotCount: ..., encapsulation: ..., styles: ...}, type: 0, viewUtils: ViewUtils{_renderer: ..., _appId: ..., _nextCompTypeId: ..., sanitizer: ...}, parentInjector: ReflectiveInjector_{_debugContext: ..., _constructionCounter: ..., _proto: ..., _parent: ..., _strategy: ...}, declarationAppElement: null, cdMode: 2, contentChildren: [], viewChildren: [], viewContainerElement: null, cdState: 2, destroyed: false, ref: ViewRef_{_view: ...}, renderer: DebugDomRenderer{_delegate: ...}, staticNodeDebugInfos: [...], _currentDebugContext: DebugContext{_view: ..., _nodeIndex: ..., _tplRow: ..., _tplCol: ...}, context: Object{}, _hasExternalHostElement: true, projectableNodes: []}, _nodeIndex: 0, _tplRow: 0, _tplCol: 0}
PhantomJS 2.1.1 (Windows 8 0.0.0) ERROR: 'EXCEPTION: Error in :0:0'

**Comment by lathonez** [comment]
Hey,

Can you check the following:

* node_modules/karma/static/context.html is identical to [test/karma-static/context.htm](https://github.com/lathonez/clicker/blob/master/test/karma-static/context.html)
* node_modules/karma/static/debug.html is identical to [test/karma-static/debug.htm](https://github.com/lathonez/clicker/blob/master/test/karma-static/debug.html)

Next steps:

* Try replacing an inline template in app.ts with an html file
* test this repo to see you have the same error

**Comment by srishiva89** [comment]
Great Thanks @lathonez. It helps me to resolve ion-app issue. But still getting below errors:-

A:\project\dx2>npm test

> Diagnox@1.0.1 test A:\project\dx2
> gulp --gulpfile test/gulpfile.ts --cwd ./ unit-test

[09:52:15] Requiring external module ts-node/register
[09:52:19] Using gulpfile A:\project\dx2\test\gulpfile.ts
[09:52:19] Starting 'unit-test'...
[09:52:19] Starting 'lint'...
[09:52:19] Starting 'html'...
[09:52:19] Finished 'html' after 569 ms
[09:52:19] Finished 'lint' after 864 ms
[09:52:19] Starting 'karma'...

START:
28 06 2016 09:52:31.879:INFO [framework.browserify]: bundle built
28 06 2016 09:52:31.905:INFO [karma]: Karma v0.13.22 server started at http://localhost:9876/
28 06 2016 09:52:31.914:INFO [launcher]: Starting browser PhantomJS
28 06 2016 09:52:33.505:INFO [PhantomJS 2.1.1 (Windows 8 0.0.0)]: Connected on socket /#7JuKpr7UBvTFoCTxAAAA with id 77815560
PhantomJS 2.1.1 (Windows 8 0.0.0) LOG: 'Angular 2 is running in the development mode. Call enableProdMode() to enable the production mode.'
  DxApp
    √ initialises with an app
    √ initialises with a root page
    √ initialises with two pages
    √ opens a page
  DashboardPage
    √ initialises
**LOG: 'labs >>> 0'**
  LabsPage
    √ initialises
LOG: 'labs >>> 0'
    √ should get a list of labs
LOG: 'labs >>> 0'
    Lab
      **× should have a name**
  Utils
    √ resets a control

Finished in 4.872 secs / 4.829 secs

SUMMARY:
√ 8 tests completed
× 1 test failed

FAILED TESTS:
  LabsPage
    Lab
28 06 2016 09:52:40.201:WARN [reporter]: **SourceMap position not found for trace: TypeError: undefined is not an object (evaluating 'lab.querySelector')** in http://localhost:9876/absoluteC:/Users/HP/AppData/Local/Temp/c907f8cea96651745fea5b03884d3062.browserify?b8c8b61ecf95693c15138a0a700fe33c7c0186ce (line 173)
http://localhost:9876/absoluteC:/Users/HP/AppData/Local/Temp/c907f8cea96651745fea5b03884d3062.browserify?b8c8b61ecf95693c15138a0a700fe33c7c0186ce:173:33
http://localhost:9876/absoluteC:/Users/HP/AppData/Local/Temp/c907f8cea96651745fea5b03884d3062.browserify?b8c8b61ecf95693c15138a0a700fe33c7c0186ce:32121:28
invoke@http://localhost:9876/base/node_modules/zone.js/dist/zone.js?8b9c8d0a7e549abe9edec8da5a095f206d206277:323:34
run@http://localhost:9876/base/node_modules/zone.js/dist/zone.js?8b9c8d0a7e549abe9edec8da5a095f206d206277:216:50
http://localhost:9876/base/node_modules/zone.js/dist/zone.js?8b9c8d0a7e549abe9edec8da5a095f206d206277:571:61
invokeTask@http://localhost:9876/base/node_modules/zone.js/dist/zone.js?8b9c8d0a7e549abe9edec8da5a095f206d206277:356:43
runTask@http://localhost:9876/base/node_modules/zone.js/dist/zone.js?8b9c8d0a7e549abe9edec8da5a095f206d206277:256:58
drainMicroTaskQueue@http://localhost:9876/base/node_modules/zone.js/dist/zone.js?8b9c8d0a7e549abe9edec8da5a095f206d206277:474:43
invoke@http://localhost:9876/base/node_modules/zone.js/dist/zone.js?8b9c8d0a7e549abe9edec8da5a095f206d206277:426:41
      × should have a name
        PhantomJS 2.1.1 (Windows 8 0.0.0)
      TypeError: undefined is not an object (evaluating 'lab.querySelector') in C:/Users/HP/AppData/Local/Temp/c907f8cea96651745fea5b03884d3062.browserify (line 173)
      C:/Users/HP/AppData/Local/Temp/c907f8cea96651745fea5b03884d3062.browserify:173:33 <- app/pages/labs/labs.spec.ts:81:17
      C:/Users/HP/AppData/Local/Temp/c907f8cea96651745fea5b03884d3062.browserify:32121:28 <- node_modules/@angular/core/testing/testing.js:98:0
      invoke@A:/project/dx2/node_modules/zone.js/dist/zone.js:323:34
      run@A:/project/dx2/node_modules/zone.js/dist/zone.js:216:50
      A:/project/dx2/node_modules/zone.js/dist/zone.js:571:61
      invokeTask@A:/project/dx2/node_modules/zone.js/dist/zone.js:356:43
      runTask@A:/project/dx2/node_modules/zone.js/dist/zone.js:256:58
      drainMicroTaskQueue@A:/project/dx2/node_modules/zone.js/dist/zone.js:474:43
      invoke@A:/project/dx2/node_modules/zone.js/dist/zone.js:426:41

------------------------------|----------|----------|----------|----------|----------------|
File                          |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
------------------------------|----------|----------|----------|----------|----------------|
 app\                         |      100 |      100 |      100 |      100 |                |
  app.ts                      |      100 |      100 |      100 |      100 |                |
 app\pages\dashboard\         |      100 |      100 |      100 |      100 |                |
  dashboard.ts                |      100 |      100 |      100 |      100 |                |
 app\pages\labs\              |      100 |      100 |      100 |      100 |                |
  labs.ts                     |      100 |      100 |      100 |      100 |                |
 app\providers\labs-provider\ |      100 |      100 |      100 |      100 |                |
  labs-provider.ts            |      100 |      100 |      100 |      100 |                |
 app\services\                |       80 |      100 |       40 |    83.33 |                |
  utils.ts                    |       80 |      100 |       40 |    83.33 |           8,10 |
------------------------------|----------|----------|----------|----------|----------------|
All files                     |     97.5 |      100 |    89.29 |    97.18 |                |
------------------------------|----------|----------|----------|----------|----------------|

[09:52:40] 'karma' errored after 21 s
[09:52:40] Error: 1
    at formatError (A:\project\dx2\node_modules\gulp\bin\gulp.js:169:10)
    at Gulp.<anonymous> (A:\project\dx2\node_modules\gulp\bin\gulp.js:195:15)
    at emitOne (events.js:82:20)
    at Gulp.emit (events.js:169:7)
    at Gulp.Orchestrator._emitTaskDone (A:\project\dx2\node_modules\gulp\node_modules\orchestrator\index.js:264:8)
    at A:\project\dx2\node_modules\gulp\node_modules\orchestrator\index.js:275:23
    at finish (A:\project\dx2\node_modules\gulp\node_modules\orchestrator\lib\runTask.js:21:8)
    at cb (A:\project\dx2\node_modules\gulp\node_modules\orchestrator\lib\runTask.js:29:3)
    at removeAllListeners (A:\project\dx2\node_modules\karma\lib\server.js:336:7)
    at Server.<anonymous> (A:\project\dx2\node_modules\karma\lib\server.js:347:9)
    at Server.g (events.js:260:16)
    at emitNone (events.js:72:20)
    at Server.emit (events.js:166:7)
    at emitCloseNT (net.js:1521:8)
    at nextTickCallbackWith1Arg (node.js:431:9)
    at process._tickCallback (node.js:353:17)
[09:52:40] 'unit-test' errored after 22 s
[09:52:40] Error in plugin 'run-sequence(karma)'
Message:
    karma callback
npm ERR! Test failed.  See above for more details.

labs.html:-

 <ion-list>
    <ion-item *ngFor="let lab of labs"> // In UI, I'm getting the data. 
      **<h3>{{lab.shortName}}</h3>**  
      <p>
        <span class="city">{{lab.city}}</span>
      </p>
    </ion-item>
  </ion-list>

labs.spec.ts:-

 beforeEach(injectAsync([TestComponentBuilder], (tcb: TestComponentBuilder) => {
    return tcb
      .createAsync(LabsPage)
      .then((componentFixture: ComponentFixture<Type>) => {
        labsPageFixture = componentFixture;
        labsPage = componentFixture.componentInstance;
        labsPageFixture.detectChanges();
        nativeElement = labsPageFixture.nativeElement;
        list = nativeElement.querySelectorAll('ion-list');
        **labs = nativeElement.querySelectorAll('ion-item');**
        console.log('labs >>> ' + labs.length);
        **lab = labs[0];** // Here i'm getting undefined.
      })
      .catch(Utils.promiseCatchHandler);
  }));

Please let me know if you need any other information on this.

**Comment by lathonez** [comment]
I usually get to the bottom of these issues by debugging in Chrome. e.g. you need to work out why your querySelectorAll('ion-item') isn't returning anything. I'd try inspecting the children of naitiveElement, it may not be what you think it is.

If you need any further help please raise a new issue.
